File: 3976-maximum-subarray-sum-after-multiplier/3976-maximum-subarray-sum-after-multiplier.py
class Solution:
    INF = float('-inf')
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        def f1()->int:
            a = float('-inf')
            b = float('-inf')
            c = float('-inf')
            maxi = float('-inf')
            for i,val in enumerate(nums):
                temp = val*k
                nexta = max(val,a+val)
                nextb = max([temp,a+temp,b+temp])
                nextc = max(b+val,c+val)
                a = nexta
                b = nextb
                c = nextc
                maxi = max([maxi,b,c]);
            return maxi
        def f2()->int:
            a = float('-inf')
            b = float('-inf')
            c = float('-inf')
            maxi = float('-inf')
            for i,val in enumerate(nums):
                temp = int(val/k)
                nexta = max(val,a+val)
                nextb = max([temp,a+temp,b+temp])
                nextc = max(b+val,c+val)
                a = nexta
                b = nextb
                c = nextc
                maxi = max([maxi,b,c]);
            return maxi
        ans1 = f1()
        ans2 = f2()
        res = float('-inf')
        curr = 0
        index =0
        while index<len(nums):
            curr = max(nums[index],curr+nums[index])
            res = max(res,curr)
            index+=1
        return max([res,ans1,ans2])

File: 3976-maximum-subarray-sum-after-multiplier/3976-maximum-subarray-sum-after-multiplier.cpp
class Solution {
public:
    long long maxSubarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        auto f = [&](bool is_multiply) -> long long {
            const long long MIN_INF = -1e16;
            long long nottaken = MIN_INF;
            long long operationcontinue = MIN_INF;
            long long finished = MIN_INF;
            long long res = MIN_INF;
            for (int i = 0; i < n; i++) {
                long long x = nums[i];
                long long op_x = is_multiply ? (x * k) : (x / k);
                long long next_nottaken = max(x, nottaken + x);
                long long next_operationcontinue = max({op_x,nottaken + op_x,operationcontinue + op_x});
                long long next_finished = max(finished + x,operationcontinue + x);
                nottaken = next_nottaken;
                operationcontinue = next_operationcontinue;
                finished = next_finished;
                res = max({res, nottaken,operationcontinue, finished});
            }
            
            return res;
        };
        long long case1 = f(true);
        long long case2 = f(false);
        
        return max(case1, case2);
    }
};